src/txmempool.cpp

```diff
@@ -693,28 +686,29 @@ void CTxMemPool::check(CChainState& active_chainstate) const
     uint64_t checkTotal = 0;
     CAmount check_total_fee{0};
     uint64_t innerUsage = 0;
+    uint64_t prev_ancestor_count{0};
 
     CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
     CCoinsViewCache mempoolDuplicate(const_cast<CCoinsViewCache*>(&active_coins_tip));
     const int64_t spendheight = active_chainstate.m_chain.Height() + 1;
 
-    std::list<const CTxMemPoolEntry*> waitingOnDependants;
-    for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
+    for (const auto& it : GetSortedDepthAndScore()) {
         unsigned int i = 0;
         checkTotal += it->GetTxSize();
         check_total_fee += it->GetFee();
         innerUsage += it->DynamicMemoryUsage();
         const CTransaction& tx = it->GetTx();
         innerUsage += memusage::DynamicUsage(it->GetMemPoolParentsConst()) + memusage::DynamicUsage(it->GetMemPoolChildrenConst());
-        bool fDependsWait = false;
         CTxMemPoolEntry::Parents setParentCheck;
         for (const CTxIn &txin : tx.vin) {
             // Check that every mempool transaction's inputs refer to available coins, or other mempool tx's.
             indexed_transaction_set::const_iterator it2 = mapTx.find(txin.prevout.hash);
             if (it2 != mapTx.end()) {
                 const CTransaction& tx2 = it2->GetTx();
                 assert(tx2.vout.size() > txin.prevout.n && !tx2.vout[txin.prevout.n].IsNull());
-                fDependsWait = true;
+                // Sanity check: all parents should have been checked before their children, which
+                // means its outputs must be available in the coins cache.
+                assert(mempoolDuplicate.HaveCoin(txin.prevout));
```

Comment:
Can this assert and the `assert(active_coins_tip.HaveCoin(txin.prevout));` in the else branch below now be merged into a single `assert(mempoolDuplicate.HaveCoin(txin.prevout));`?

If I understand correctly, the `mempoolDuplicate` coins cache is a copy of the `active_chains_tip` cache, so should have all of the coins in the UTXO set. That means `mempoolDuplicate` should have all of the coins that are being spent by each transaction as we iterate through the entries in depth order.

```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -706,13 +706,11 @@ void CTxMemPool::check(CChainState& active_chainstate) const
             if (it2 != mapTx.end()) {
                 const CTransaction& tx2 = it2->GetTx();
                 assert(tx2.vout.size() > txin.prevout.n && !tx2.vout[txin.prevout.n].IsNull());
-                // Sanity check: all parents should have been checked before their children, which
-                // means its outputs must be available in the coins cache.
-                assert(mempoolDuplicate.HaveCoin(txin.prevout));
                 setParentCheck.insert(*it2);
-            } else {
-                assert(active_coins_tip.HaveCoin(txin.prevout));
             }
+            // Sanity check: all parents should have been checked before their children, which
+            // means its outputs must be available in the coins cache.
+            assert(mempoolDuplicate.HaveCoin(txin.prevout));
             // Check whether its inputs are marked in mapNextTx.
             auto it3 = mapNextTx.find(txin.prevout);
             assert(it3 != mapNextTx.end());
```

Reply:
Yes, it can, but I thought people might find it scary. I've grouped this and the !IsCoinbase check you suggested into one "simplify logic" commit.